File: mobile/packages/divine_camera/android/src/main/kotlin/co/openvine/divine_camera/DivineCameraLog.kt
// ABOUTME: Forwards curated native camera diagnostics to Dart's UnifiedLogger
// ABOUTME: so recording issues (e.g. missing audio) appear in user bug reports

package co.openvine.divine_camera

import android.util.Log

/**
 * Bridges relevant native diagnostics to the Dart side so they are captured by
 * the app's `UnifiedLogger` and included in bug-report log dumps.
 *
 * Native recording code calls these methods (instead of bare [Log]) for the
 * handful of events worth surfacing to support: audio-permission denial,
 * recording finalize errors, encoder-fallback retries, and audio-source state
 * transitions. Per-frame / verbose logging must stay on [Log] so it does not
 * flood the captured buffer.
 *
 * `DivineCameraPlugin` installs [sink] when attached to the engine to forward
 * each entry over the method channel. Before that is wired (or in unit-test
 * contexts) entries fall back to logcat only.
 */
object DivineCameraLog {
    /**
     * Forwards `(level, message, name)` to Dart. `level` is one of `debug`,
     * `info`, `warning`, `error`. Set by the plugin; `null` until then.
     */
    @Volatile
    var sink: ((String, String, String) -> Unit)? = null

    fun debug(message: String, name: String = "DivineCamera") = emit("debug", message, name)

    fun info(message: String, name: String = "DivineCamera") = emit("info", message, name)

    fun warning(message: String, name: String = "DivineCamera") = emit("warning", message, name)

    fun error(message: String, name: String = "DivineCamera") = emit("error", message, name)

    // android.util.Log-shaped overloads so existing `Log.x(TAG, ...)` call
    // sites forward to the bridge with a mechanical rename. The tag becomes
    // the entry name. Logcat keeps the full throwable stack trace; Dart gets a
    // concise throwable summary because MethodChannel payloads stay structured.
    fun d(tag: String, message: String, tr: Throwable? = null) =
        emit("debug", message, tag, tr)

    fun i(tag: String, message: String, tr: Throwable? = null) =
        emit("info", message, tag, tr)

    fun w(tag: String, message: String, tr: Throwable? = null) =
        emit("warning", message, tag, tr)

    fun e(tag: String, message: String, tr: Throwable? = null) =
        emit("error", message, tag, tr)

    private fun emit(level: String, message: String, name: String, tr: Throwable? = null) {
        // Keep the logcat fallback so on-device debugging is unchanged.
        when (level) {
            "error" -> if (tr != null) Log.e(name, message, tr) else Log.e(name, message)
            "warning" -> if (tr != null) Log.w(name, message, tr) else Log.w(name, message)
            "info" -> if (tr != null) Log.i(name, message, tr) else Log.i(name, message)
            else -> if (tr != null) Log.d(name, message, tr) else Log.d(name, message)
        }
        sink?.invoke(level, tr?.let { "$message: $it" } ?: message, name)
    }
}

File: mobile/packages/divine_camera/ios/Classes/DivineCameraLog.swift
// ABOUTME: Forwards curated native camera diagnostics to Dart's UnifiedLogger
// ABOUTME: so recording issues (e.g. missing audio) appear in user bug reports

import Foundation

/// Bridges relevant native diagnostics to the Dart side so they are captured
/// by the app's `UnifiedLogger` and included in bug-report log dumps.
///
/// Native recording code calls these methods (instead of bare `print`) for the
/// handful of events worth surfacing to support: audio-session configuration,
/// interruptions and recovery, recording start/stop, and asset-writer
/// failures. Per-frame / verbose logging must stay on `print` so it does not
/// flood the captured buffer.
///
/// `DivineCameraPlugin` installs `sink` at registration to forward each entry
/// over the method channel. Before that is wired (or in unit/host contexts)
/// entries fall back to the console only.
final class DivineCameraLog {
    static let shared = DivineCameraLog()

    private init() {}

    /// Forwards `(level, message, name)` to Dart. `level` is one of
    /// `debug`, `info`, `warning`, `error`. Set by the plugin; `nil` until then.
    var sink: ((String, String, String) -> Void)?

    func debug(_ message: String, name: String = "DivineCamera") {
        emit("debug", message, name)
    }

    func info(_ message: String, name: String = "DivineCamera") {
        emit("info", message, name)
    }

    func warning(_ message: String, name: String = "DivineCamera") {
        emit("warning", message, name)
    }

    func error(_ message: String, name: String = "DivineCamera") {
        emit("error", message, name)
    }

    private func emit(_ level: String, _ message: String, _ name: String) {
        // Keep the console fallback so on-device debugging is unchanged.
        print("[\(name)] \(message)")
        sink?(level, message, name)
    }
}
